packages/server-wallet/src/protocols/open-channel.ts

```diff
@@ -46,304 +32,93 @@ export class ChannelOpener {
-              case 'SignState':
-                await this.signState(nextAction, protocolState, tx, response);
-                break;
-              case 'FundChannel':
-                await this.fundChannel(nextAction, tx);
-                break;
-              case 'CompleteObjective':
-                attemptAnotherProtocolStep = false;
-                await this.completeObjective(objective, protocolState, tx, response);
-                break;
-              case 'RequestLedgerFunding':
-                await this.requestLedgerFunding(protocolState, tx);
-                break;
-              default:
-                unreachable(nextAction);
-            }
-          } catch (error) {
-            this.logger.error({error}, 'Error handling action');
-            await tx.rollback(error);
-            attemptAnotherProtocolStep = false;
-          }
-        } else {
-          response.queueChannelState(protocolState.app);
-          attemptAnotherProtocolStep = false;
-        }
-      });
-    }
-  }
+      // if we haven't signed the pre-fund, sign it
+      if (!channel.prefundSigned) {
+        await this.signPrefundSetup(channel, response, tx);
+      }
 
-  private async signState(
-    action: SignState,
-    protocolState: ProtocolState,
-    tx: Transaction,
-    response: WalletResponse
-  ): Promise<void> {
-    const {myIndex, channelId} = protocolState.app;
-    const channel = await Channel.forId(channelId, tx);
-    const signedState = await this.store.signState(channel, action, tx);
-    response.queueState(signedState, myIndex, channelId);
-  }
+      // if we don't have a supported preFundSetup, we're done for now
+      if (!channel.prefundSupported) {
+        return;
+      }
 
-  private async fundChannel(action: FundChannel, tx: Transaction): Promise<void> {
-    await ChainServiceRequest.insertOrUpdate(action.channelId, 'fund', tx);
-    await this.chainService.fundChannel({
-      ...action,
-      expectedHeld: BN.from(action.expectedHeld),
-      amount: BN.from(action.amount),
-    });
-  }
+      // if we have a full pre fund setup, delegate to the funding process to (a) see if
+      // the channel is funded, and (b) take action if not
+      const funded = await this.crankChannelFunder(objective, channel, response, tx);
 
-  private async completeObjective(
-    objective: DBOpenChannelObjective,
-    protocolState: ProtocolState,
-    tx: Transaction,
-    response: WalletResponse
-  ): Promise<void> {
-    await this.store.markObjectiveAsSucceeded(objective, tx);
+      // if the channel is funded and I haven't signed the post-fund, sign it
+      if (funded && !channel.postfundSigned) {
+        await this.signPostfundSetup(channel, response, tx);
+      }
 
-    response.queueChannelState(protocolState.app);
-    response.queueSucceededObjective(objective);
-  }
+      if (channel.postfundSupported) {
+        await this.store.markObjectiveAsSucceeded(objective, tx);
```

Comment:
See my other comment about possibly returning `true` as an alternative way to signal this? That way, we can also return `false` in irrevocable failure.

Reply:
I think that this is doing something different - it's updating the store, so that no future runs need to visit this objective, and the objective could even be removed. There's not really anything calling the `ChannelOpener.crank` that is going to change its behaviour base on the result - other than send slightly different messages to the user and opponent.

Reply:
I guess my thought was that the `ObjectiveManager` (the thing calling `ChannelOpener.crank`) could interpret the return value and then mark the objective as completed in the store (if appropriate). That way, you could have a uniform interface for all objects that can be cranked, which might aid in composing them.